openshift/machine-config-operator
# Send alert when MCO can't safely apply updated Kubelet CA on nodes in paused pool

**- What I did**

1. Added metrics reporting to machine-config-controller because it did not previously have that capability by adding:  
- In manifests:
   - Cluster Roles
   - Cluster Role Bindings
   - ServiceMonitor for metrics
   - Service for metrics
   - oauth-proxy container for machine-config-controller deployment
   - mcc-proxy-tls secret for machine-config-controller
  
- In controller:
   - metrics handler function in machine-config-controller common
   - machine config lister in node_controller and node_controller_test
- References:
   - For the handler I cribbed off of: 557303ffd70ba6ccfa834f86ff7d4acd1f97dd17
   - And then to add oauth: 3ab692f423cb34d70d7cb4dc808206849362ba05	

2. Added an alert that fires when the kubelect-ca certificate is pending in a paused pool: 
   - added a GaugeVec (MCCImportantConfigPaused) for important pending config 
   - added a map in node_controller to store the table of "important config names" : "important config files"
   - added functions to check for pending files, and release the alerts once the pools have been unpaused
 
3. I don't have tests yet

**- How to verify it**

1. To fire the alert:

   - build a cluster
   - oc edit mcp worker 
   - change spec.paused: false to spec.paused: true 
   - Trigger a certificate rotation:  ```oc patch secret -p='{"metadata": {"annotations": {"auth.openshift.io/certificate-not-after": null}}}' kube-apiserver-to-kubelet-signer -n openshift-kube-apiserver-operator```
   - Observe prometheus metrics: ```TOKEN=`oc sa get-token prometheus-k8s -n openshift-monitoring`  
POD=`oc get pods -n openshift-machine-config-operator | grep controller | awk '{print $1}'`  
oc rsh $POD curl -k -H "Authorization: Bearer $TOKEN" https://localhost:9001/metrics  ```
   - look for mcc_important_config_paused
   - (or watch the web UI, they show up there too)
   
 2. To stop the alert: 
    - oc edit mcp worker 
    - change spec.paused: true to spec.paused: false 
    - once again observe prometheus metrics -- the alert will stop firing/set to 0 

**- Description for the changelog**
Send alert when MCO can't safely apply updated Kubelet CA on nodes in paused pool
